### PR TITLE
[ADD] report_page_orientation

### PR DESCRIPTION
**Pull request by KKamaa**
So having issue with pre-commit due to click version as described [here ](https://github.com/dask/distributed/issues/6013) an proper version is required.

### CODE REVIEW

**Comment by thomaspaulb** [comment]
@NL66278 This is the current PR including tests

File: report_page_orientation/views/views.xml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    <template id="minimal_layout" inherit_id="web.external_layout_standard">
+        <xpath expr="//span[@class='page']" position="after">
+            <span><t t-esc="env.context.get('num_page')" /></span>
+         </xpath>
+        <xpath expr="//span[@class='topage']" position="after">
+            <span><t t-esc="env.context.get('tot_page')"/></span>
+         </xpath>
+     </template>
+     -->
+</odoo>
```
Comment (thomaspaulb) [inline report_page_orientation/views/views.xml]: Can we remove Gio's abandoned great ideas?

File: report_wkhtmltopdf_param/tests/test_report_paperformat.py [hunk 2]
```diff
@@ -23,6 +36,8 @@ def test_wkhtmltopdf_valid_parameter(self):
                 report_paperformat.update(
                     {"custom_params": [(0, 0, {"name": "--disable-smart-shrinking"})]}
                 )
+            except ParserError:
+                _logger.error("Document is empty")
```
Comment (thomaspaulb) [inline report_wkhtmltopdf_param/tests/test_report_paperformat.py]: Hey this is a change to another module, did this come from you in order to make it green? Or from elsewhere

**Review by thomaspaulb — CHANGES_REQUESTED** [review]

File: report_wkhtmltopdf_param/tests/test_report_paperformat.py (hunk 2 above)
Comment (KKamaa) [reply report_wkhtmltopdf_param/tests/test_report_paperformat.py]: Yeah did this to make it green.

Comment (thomaspaulb) [reply report_wkhtmltopdf_param/tests/test_report_paperformat.py]: Ok but then in which line does it throw ParserError when you run the tests?

https://github.com/OCA/reporting-engine/blob/15.0/report_wkhtmltopdf_param/models/report_paperformat.py#L41

Comment (KKamaa) [reply report_wkhtmltopdf_param/tests/test_report_paperformat.py]: so I was due to a _**"Document is Empty"**_ error this can be seen at [line 306](https://github.com/OCA/reporting-engine/runs/5760772752?check_suite_focus=true#step:8:306) from the previous build, that is  why I added the exception part. Mostly on a new demo db.

Comment (thomaspaulb) [reply report_wkhtmltopdf_param/tests/test_report_paperformat.py]: Actually, something else was going on. In the [15.0 migration](https://github.com/OCA/reporting-engine/pull/579/files) in [this commit](https://github.com/OCA/reporting-engine/commit/3638d7d0b5510da400334bfa9b0ab18a3f615197) the `content_html` array was replaced by `sample_html`, and no-one complained about it during review, maybe it was overlooked.

As a result, `_run_wkhtmltopdf` was run directly with a HTML string as the second parameter, instead of a `bodies` array of strings like expected.

So `for i, body in enumerate(bodies)` splits up the string into its 293 characters which miraculously does not lead to problems in the normal test, but when we added our code, we are overriding `_run_wkhtmltopdf` and actually trying to decode each of the bodies from HTML to a structure, and then it gives an error.

The correct solution would thus be not to change the test to deal with ParserError, but to prevent ParserError in the first place by reverting the change that was made in the module.

File: report_wkhtmltopdf_param/models/report_paperformat.py [hunk 3]
```diff
@@ -36,7 +36,8 @@ def _check_recursion(self):
             report = self.env["ir.actions.report"].new(
                 {"paperformat_id": paperformat.id}
             )
-            content = report._run_wkhtmltopdf(sample_html)
+            contenthtml = [bytes(sample_html, "utf-8")]
```
Comment (thomaspaulb) [inline report_wkhtmltopdf_param/models/report_paperformat.py]: Test now fails on the `bytes` object.... maybe it should be just `[sample_html]` ?

Comment (KKamaa) [reply report_wkhtmltopdf_param/models/report_paperformat.py]: @thomaspaulb think we good now :smiling_face_with_tear:

Comment (thomaspaulb) [reply report_wkhtmltopdf_param/models/report_paperformat.py]: Yayyy green

**Review by thomaspaulb — APPROVED** [review]

File: report_page_orientation/demo/report_page_orientation_demo.xml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            <t t-call="web.external_layout">
+                <div class="page">
+                    List of Current Users in The system
+                    <ul>
+                        <li t-foreach="docs" t-as="doc">
+                            <t t-esc="doc.name" />
+                        </li>
+                    </ul>
+                    <h2>Hello this is a page</h2>
+                    <t t-foreach="range(100)" t-as="index">
```
Comment (gfcapalbo) [inline report_page_orientation/demo/report_page_orientation_demo.xml:29]: @tom this does one portrait page (1xp), 100 rotated pages (100 x r) and closes with a portrait page (1x p)  -  the 100 number seems unnecessary.
wouldn't it be better to have test data that checks different patterns? ( different alternations, with different first pages and last pages, to uncover corner cases) 
1x p - 3 x r - 1x r
1x r - 3x p - 2 x r 


the test infrastructure is there should be easy to adapt.

File: report_page_orientation/models/ir_actions_report.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
+# Copyright 2022 Sunflower IT <https://www.sunflowerweb.nl>.
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+import io
+import logging
+import tempfile
+from copy import deepcopy
+
```
Comment (gfcapalbo) [inline report_page_orientation/models/ir_actions_report.py:8]: why space?

File: report_page_orientation/models/ir_actions_report.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+                if num_pages > 1:
+                    for _i in range(num_pages - 1):
+                        page_break.addprevious(deepcopy(morepage))
+            if sections:
+                body = lxml.html.tostring(root, encoding="unicode")
+            new_bodies.append(body)
+
+        # If no change, just return super.
+        if not bodies_with_rotation:
+            return super(IrActionsReport, self)._run_wkhtmltopdf(
```
Comment (gfcapalbo) [inline report_page_orientation/models/ir_actions_report.py:123]: odoo 15/ python 3X  we can write super()._run_wkhtmltopdf()  
"old" nomenclature is necessary if you need a modified "self" e.g. self.with_context()  or self.sudo()
or whatever. but in this case the more concise version is Okay.

File: report_page_orientation/README.rst [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+Authors
+~~~~~~~
+
+* Sunflower IT
+
+Contributors
+~~~~~~~~~~~~
+
+* Giovanni Capalbo <giovanni@therp.nl>
```
Comment (gfcapalbo) [inline report_page_orientation/README.rst:73]: @thomaspaulb i should not be first author. 
made core idea , but the "placeholder" substitution, the wkhtml2pdf improvements, AND most importantly the the JS subst solution  are yours. keep alphabetical IMO.

File: report_page_orientation/models/ir_actions_report.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+            if (section_counter < len(sections)) and sections[section_counter].get(
+                "page", 0
+            ) == input_page_counter + 1:
+                footer_element = lxml.html.fromstring(footer)
+                subst_elements = footer_element.xpath(
+                    "//script[contains(text(), 'subst')]"
+                )
+                multi_footer = footer
+                if subst_elements:
+                    subst2 = lxml.html.fragment_fromstring(
```
Comment (gfcapalbo) [inline report_page_orientation/models/ir_actions_report.py:172]: killer trick.

**Review by gfcapalbo — APPROVED** [review]
LGTM, some minor observations. :+1:

File: report_page_orientation/views/views.xml (hunk 1 above)
Comment (gfcapalbo) [reply report_page_orientation/views/views.xml]: @thomaspaulb     a  "great idea" can be removed when a working idea exists.
removing this myself.

Comment (gfcapalbo) [reply report_page_orientation/views/views.xml]: i see it was cleaned up. ignore.

File: report_page_orientation/README.rst (hunk 7 above)
Comment (thomaspaulb) [reply report_page_orientation/README.rst:73]: @KKamaa Could you fix this and add the both of us too, and then squash into a single commit? After that it is ready for merge.

Comment (KKamaa) [reply report_page_orientation/README.rst:73]: @thomaspaulb added the contributors though kept it alphabetical and squashed to one commit as you mentioned

File: report_page_orientation/models/ir_actions_report.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+                    )
+                    subst_elements[-1].addnext(subst2)
+                    multi_footer = lxml.html.tostring(
+                        footer_element, encoding="unicode"
+                    )
+
+                section = sections[section_counter]
+                num_pages = section["num_pages"]
+                section_pdf_content = super(IrActionsReport, self)._run_wkhtmltopdf(
+                    [section["content"]],
```
Comment (thomaspaulb) [inline report_page_orientation/models/ir_actions_report.py:212]: @gfcapalbo and me are currently investigating a case whereby relevant CSS was not included in the landscaped version, which would then be because the HTML included here does not contain any wrapping. Pending this investigation, this should not be merged.

**Review by thomaspaulb — CHANGES_REQUESTED** [review]
